File: files-app/src/main/java/l/files/operations/ui/FailureMessage.java
package l.files.operations.ui;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.auto.value.AutoValue;

import l.files.fs.File;

@AutoValue
abstract class FailureMessage implements Parcelable {

    FailureMessage() {
    }

    abstract File resource();

    abstract String message();

    static FailureMessage create(File file, String message) {
        return new AutoValue_FailureMessage(file, message);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(resource(), 0);
        dest.writeString(message());
    }

    public static final Creator<FailureMessage> CREATOR = new Creator<FailureMessage>() {

        @Override
        public FailureMessage createFromParcel(Parcel source) {
            File file = source.readParcelable(FailureMessage.class.getClassLoader());
            String message = source.readString();
            return create(file, message);
        }

        @Override
        public FailureMessage[] newArray(int size) {
            return new FailureMessage[size];
        }

    };

}

File: files-app/src/main/kotlin/operations/ui/FailureMessage.kt
package l.files.operations.ui

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import l.files.fs.Path

data class FailureMessage(val path: Path, val message: String) : Parcelable {

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(path, 0)
        dest.writeString(message)
    }

    override fun describeContents() = 0

    class object {

        val CREATOR: Creator<FailureMessage> = object : Creator<FailureMessage> {
            override fun createFromParcel(source: Parcel): FailureMessage {
                val path = source.readParcelable<Path>(javaClass.getClassLoader())
                val message = source.readString()
                return FailureMessage(path, message)
            }

            override fun newArray(size: Int) = arrayOfNulls<FailureMessage>(size)
        }

    }

}
